bootstrap/internal/controllers/rke2config_controller.go

```diff
@@ -225,25 +240,329 @@ func (r *Rke2ConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	scope.Logger.V(5).Info("Showing marshalled config.yaml", "config.yaml", string(b))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	scope.Logger.Info("Joining Worker config marshalled successfully")
+
+	wkJoinConfigFile := bootstrapv1.File{
+		Path:        rke2.DefaultRKE2ConfigLocation,
+		Content:     string(b),
+		Owner:       "root:root",
+		Permissions: "0640",
+	}
+
+	// TODO: Implement adding additional files through API
+
+	wkInput :=
+		&cloudinit.BaseUserData{
+			PreRKE2Commands:  scope.Config.Spec.PreRKE2Commands,
+			AirGapped:        scope.Config.Spec.AgentConfig.AirGapped,
+			PostRKE2Commands: scope.Config.Spec.PostRKE2Commands,
+			ConfigFile:       wkJoinConfigFile,
+			RKE2Version:      scope.Config.Spec.AgentConfig.Version,
+			WriteFiles:       files,
+		}
+
+	cloudInitData, err := cloudinit.NewJoinWorker(wkInput)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.storeBootstrapData(ctx, scope, cloudInitData); err != nil {
+		return ctrl.Result{}, err
+	}
+
 	return ctrl.Result{}, nil
 }
+
+func (r *RKE2ConfigReconciler) generateAndStoreToken(ctx context.Context, scope *Scope) (string, error) {
+	tokn, err := bsutil.Random(16)
+	if err != nil {
+		return "", err
+	}
+
+	scope.Logger = scope.Logger.WithValues("cluster-name", scope.Cluster.Name)
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      bsutil.TokenName(scope.Cluster.Name),
+			Namespace: scope.Config.Namespace,
+			Labels: map[string]string{
+				clusterv1.ClusterLabelName: scope.Cluster.Name,
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: scope.Cluster.APIVersion,
+					Kind:       scope.Cluster.Kind,
+					Name:       scope.Cluster.Name,
+					UID:        scope.Cluster.UID,
+					Controller: pointer.BoolPtr(true),
+				},
+			},
+		},
+		Data: map[string][]byte{
+			"value": []byte(tokn),
+		},
+		Type: clusterv1.ClusterSecretType,
+	}
+
+	// as secret creation and scope.Config status patch are not atomic operations
+	// it is possible that secret creation happens but the config.Status patches are not applied
+	if err := r.Client.Create(ctx, secret); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return "", errors.Wrapf(err, "failed to create token for RKE2Config %s/%s", scope.Config.Namespace, scope.Config.Name)
+		}
+		// r.Log.Info("bootstrap data secret for RKE2Config already exists, updating", "secret", secret.Name, "RKE2Config", scope.Config.Name)
+		if err := r.Client.Update(ctx, secret); err != nil {
+			return "", errors.Wrapf(err, "failed to update bootstrap token secret for RKE2Config %s/%s", scope.Config.Namespace, scope.Config.Name)
+		}
+	}
+
+	return tokn, nil
+}
+
+// storeBootstrapData creates a new secret with the data passed in as input,
+// sets the reference in the configuration status and ready to true.
+func (r *RKE2ConfigReconciler) storeBootstrapData(ctx context.Context, scope *Scope, data []byte) error {
+	secret := &corev1.Secret{
```

Comment:
Seems like the code is very similar to the function above

Reply:
It is!
I moved the second part to another func (create or update logic). However, the secret object creation is a bit more specific, and refactoring introduces an unnecessary level of complexity.